cpp/include/kvikio/file_handle.hpp

```diff
@@ -107,120 +126,121 @@ inline int open_fd(const std::string& file_path, const std::string& flags, mode_
  */
 class FileHandle {
  private:
-  int _fd{-1};
-  bool _own_fd{false};
-  bool _closed{true};
+  // We use two file descriptors, one opened with the O_DIRECT flag and one without.
+  int _fd_direct_on{-1};
+  int _fd_direct_off{-1};
+  bool _initialized{false};
+  bool _compat_mode{false};
   mutable std::size_t _nbytes{0};  // The size of the underlying file, zero means unknown.
   CUfileHandle_t _handle{};
 
  public:
   static constexpr mode_t m644 = S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH;
-  FileHandle()                 = default;
+  FileHandle() noexcept        = default;
 
   /**
-   * @brief Construct a file handle from an existing file decriptor
+   * @brief Construct a file handle from a file path
    *
-   * @param fd File decriptor
-   * @param steal_fd When true, the handle owns the file descriptor and will close it
-   * on destruction.
+   * FileHandle opens the file twice and maintains two file descriptors. One file are
+   * opened with the specified `flags` and the other file are opened with the `flags`
+   * plus the `O_DIRECT` flag.
+   *
+   * @param file_path File path to the file
+   * @param flags Open flags (see also `fopen(3)`):
+   *   "r" -> "open for reading (default)"
+   *   "w" -> "open for writing, truncating the file first"
+   *   "a" -> "open for writing, appending to the end of file if it exists"
+   *   "+" -> "open for updating (reading and writing)"
+   * @param mode Access modes (see `open(2)`).
+   * @param compat_mode Enable KvikIO's compatibility mode for this file.
    */
-  FileHandle(int fd, bool steal_fd = false) : _fd{fd}, _own_fd{steal_fd}, _closed{false}
+  FileHandle(const std::string& file_path,
+             const std::string& flags = "r",
+             mode_t mode              = m644,
+             bool compat_mode         = defaults::compat_mode())
+    : _fd_direct_on{open_fd(file_path, flags, true, mode)},
+      _fd_direct_off{open_fd(file_path, flags, false, mode)},
```

Comment:
Why are we opening with both `O_DIRECT` and without?

Reply:
`O_DIRECT` requires page-aligned IO, which we cannot guaranty. cuFIle handles this automatically by opening the file without `O_DIRECT` as well. However, `posix_io` does not support `O_DIRECT`.

Reply:
Right, but is there a need to open with both modes? Could we for example try `O_DIRECT` and fallback to not using it (if `O_DIRECT` won't work)?

Reply:
Good point. I did this because the posix backend will need both with and without `O_DIRECT` in the future for better performance. Now, we could open the extra fd lazily but that will require that the file path is still valid (not moved or removed) at a later time. 

I think we should stick to opening two file descriptor for now, then later we can look at a more lazy approach.

Reply:
Actually, changed it so we only opens one fd when in regular mode since cuFile requires `O_DIRECT` always.